hpy/devel/src/runtime/ctx_type.c

```diff
@@ -420,6 +420,10 @@ _HPy_HIDDEN HPy
 ctx_New(HPyContext ctx, HPy h_type, void **data)
 {
     PyObject *tp = _h2py(h_type);
+    if (tp == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "HPy_New arg 1 must not be a null handle");
```

Comment:
I'm not sure this is the correct way to go:
1. CPython never (or almost never? I don't know) does these kind of checks: it just assumes that the `PyObject*` you pass around are not-NULL

2. The HPy style is to DO these checks, but in debug mode only (debug mode which doesn't exist yet, but that's another story), to not have impact on performances. I'm in favor of saying that it is responsibility of the caller to ensure that the `HPy` handles you pass around are non-NULL. 

3. What about all the other functions? Do you want to introduce NULL checks on *all* of them?

Reply:
I think the situation is a bit more complex:

1. Internally HPy mostly passes the result of `_h2py` directly to CPython functions, and assume they do something sane.  In these cases HPy is doing a bit more with the result. If we're not going to do the NULL check, why bother with the PyType_Check either? Why not just pass the result directly to `PyObject_New` and let it sort it out?

2. HPy is already a little inconsistent in places. E.g. `ctx_Dup` does XINCREF, but `ctx_FromPyObject` and `ctx_AsPyObject` do INCREF.

3. We do also have HPy_IsNull, so it's definitely expected that handles can contain null objects in some places.

Happy for you to make the call here though -- my commits were mostly so that we could have this conversation.